File: algorithms/005.longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        // 01. set the mid point
        if (s == "")    return "";
        int n = s.size(), max_strlen = 1;
        string max_str = s.substr(0,1);
        for (int mid=0; mid < n; mid++){
            int max_len = min(mid, n - 1 - mid);
            // 02-1. max odd mid palindrome
            for (int j=0; j <= max_len; j++){
                if (s[mid - j] != s[mid + j])   break;
                else if (1 + 2*j > max_strlen){
                    max_strlen = 1 + 2*j;
                    max_str = s.substr(mid - j, 1 + 2*j);
                }
            }
            
            // 02-2. max even mid palindrome
            max_len = min(mid, n - 2 - mid);
            for (int j=0; j <= max_len; j++){
                if (s[mid - j] != s[mid + 1 + j])   break;
                else if (2 + 2*j > max_strlen){
                    max_strlen = 2 + 2*j;
                    max_str = s.substr(mid - j, 2 + 2*j);
                }
            }
        }
        return max_str;
    }
};

File: algorithms/005.longest-palindromic-substring.java
class Solution {
  // O(n^2) solution for finding the middle of palindromes
	public String longestPalindrome(String s) {
		if (s == null || s.isEmpty()) return "";

		int n = s.length();
		int maxLen = 1;
		String maxStr = s.substring(0, 1);

		for (int mid = 0; mid < n; mid++) {
			// 1. Max odd-length palindrome centered at mid
			for (int j = 0; mid - j >= 0 && mid + j < n; j++) {
				if (s.charAt(mid - j) != s.charAt(mid + j)) break;
				int len = 2 * j + 1;
				if (len > maxLen) {
					maxLen = len;
					maxStr = s.substring(mid - j, mid + j + 1);
				}
			}

			// 2. Max even-length palindrome centered at mid and mid+1
			for (int j = 0; mid - j >= 0 && mid + 1 + j < n; j++) {
				if (s.charAt(mid - j) != s.charAt(mid + 1 + j)) break;
				int len = 2 * j + 2;
				if (len > maxLen) {
					maxLen = len;
					maxStr = s.substring(mid - j, mid + 1 + j + 1);
				}
			}
		}
		return maxStr;
	}
}
